### PR TITLE
Allow PDF uploads in image upload

### PR DESCRIPTION
**Pull request by caesar2164**
- expand code to allow pdf
- add lists of mime types to python and JS
- adjust js tests to be stricter and test pdf as well
- fix variable call for file uploads and studio text
- also add content type checking

### CODE REVIEW

**Comment by sarina** [comment]
Happy 2015 @caesar2164 ! This is a feature we very much want. Let me know how I can help you with it. It also looks like you haven't contributed to this repo before - so you should add yourself to AUTHORS. Feel free to do so in a separate pull request so we don't have to deal w/ conflicts in that file.

**Comment by caesar2164** [comment]
@sarina - I think this is very close to being done, but I haven't done the big job of converting all the old tests quite yet. What you could to help is let me know how this feature should support old/deprecated config. Meaning, how should the old "allow_file_upload" setting convert to the new "file_upload_type" setting? And how should we test that kind of thing...

**Comment by caesar2164** [comment]
Also, happy 2015 to you too, @sarina!

**Comment by sarina** [comment]
@caesar2164 that's a good question, that I certainly don't know how to answer! I'll see if I can get you some advice.

**Comment by sarina** [comment]
At a glance it seems like "allow_file_upload" should convert to "file_upload_type": "image". But I am unsure how to advise the testing of this.

**Comment by stephensanchez** [comment]
"allow_file_upload" is a field on the ORA2 XBlock, and it should be backwards compatible for existing instances of the block. "allow_file_upload" should automatically assume "images" for the file type, since it is the only supported file type(s) at this time. I'd recommend making a dynamic property on the XBlock to check for the "allow_file_upload" field if nothing is set for "file_upload_type". 

Testing should not be too difficult. There are entire XML-based XBlock configurations that can be loaded into a unit test, as well as a set of configurations used for the XBlock Workbench. Make sure there is an example for both, which allows unit testing of the backwards compatability, as well as testing it manually in Workbench. 

You'll also need to test this in Studio, and ensure that the "file_upload_type" parameter gets set when an old problem is re-authored. This can be done fairly easily by setting up an ORA2 problem on a sandbox, then applying your changes. If you want to keep it for future use, simply export the course and import it as needed.

Let me know if you need me to track down testing examples.

**Comment by sarina** [comment]
@caesar2164 how's this going? Need help with the testing or getting your builds to pass?

**Comment by caesar2164** [comment]
@sarina - moving along slowly...

I;d love for you to quickly see if you can figure out why tests are failing! :)

**Comment by sarina** [comment]
It looks like there's a failure to install a package: 
```
pip install -q -r requirements/test.txt
  Could not find a tag or branch 'd175715e0fe3367ec0f1ee429c242d603f6e8b10', assuming commit.
  Could not find a tag or branch '56f048af9b6868613c14aeae760548834c495011', assuming commit.
  Could not find a tag or branch '44d66a1091947fda130cb3e36135f6dfc3b135f0', assuming commit.
  Hash of the package https://pypi.python.org/packages/source/W/Werkzeug/Werkzeug-0.9.6.tar.gz#md5=f7afcadc03b0f2267bdc156c34586043 (from https://pypi.python.org/simple/werkzeug/) (34a15321b35ac2ce4d6652f6487329f5) doesn't match the expected hash f7afcadc03b0f2267bdc156c34586043!
  Bad md5 hash for package https://pypi.python.org/packages/source/W/Werkzeug/Werkzeug-0.9.6.tar.gz#md5=f7afcadc03b0f2267bdc156c34586043 (from https://pypi.python.org/simple/werkzeug/)
make: *** [install-test] Error 1
```

Have you seen this before? Can you rebase or add another commit to trigger a rebuild to see if it's a transient error?

**Comment by caesar2164** [comment]
@sarina - i haven;t seen that before. Rebased, let's see if it happens again...

**Comment by sarina** [comment]
@caesar2164 looks like that issue was fixed. However looking at your Travis output... there's a lot of failure there. Can you replicate the failures locally? I admit I don't' know much about ORA2. Can someone at Stanford help you with the tests directly? It would be faster if you could do that, because the ORA2 maintainers are the Pakistani team, so it's going to be a lot of back and forth given your time zone difference to get help from them.

**Comment by caesar2164** [comment]
@sarina - the one failing test isn't because of my commit, I think...

**Comment by stephensanchez** [comment]
@caesar2164 It looks like that test was added with the "cancel submission" feature. I just took a quick look at master and it seems like tests are passing, so you may want to try and rebase to see if there is still an issue.

**Comment by sarina** [comment]
@caesar2164 - should the title of this still be WIP? If so, what is still in progress?

**Comment by caesar2164** [comment]
@sarina & @stephensanchez - tests passed!
I still want to put this up on a sandbox to test it, but can you guys start code reviewing this now? Thanks!

**Comment by caesar2164** [comment]
@sarina - fixed!

**Comment by sarina** [comment]
@caesar2164 someone from the open source team will pick up review of this asap, and then we'll throw it to the ORA2 team, which Steve is no longer on

File: openassessment/templates/openassessmentblock/edit/oa_edit.html [hunk 1]
```diff
@@ -99,15 +99,16 @@ <h2 class="openassessment_alert_title">{% trans "Rubric Change Impacts Settings
...
-                <li id="openassessment_submission_image_wrapper" class="field comp-setting-entry">
+                <li id="openassessment_submission_file_wrapper" class="field comp-setting-entry">
                     <div class="wrapper-comp-setting">
-                        <label for="openassessment_submission_image_editor" class="setting-label">{% trans "Allow Image Responses"%}</label>
-                        <select id="openassessment_submission_image_editor" class="input setting-input" name="image submission">
-                            <option value="0">{% trans "False"%}</option>
-                            <option value="1" {% if allow_file_upload %} selected="true" {% endif %}>{% trans "True"%}</option>
+                        <label for="openassessment_submission_upload_selector" class="setting-label">{% trans "Allow File Upload"%}</label>
+                        <select id="openassessment_submission_upload_selector" class="input setting-input" name="upload submission">
+                            <option value="none">{% trans "None"%}</option>
```
Comment (sarina) [inline openassessment/templates/openassessmentblock/edit/oa_edit.html]: Can we make this clearer? Maybe, "No File Uploads Allowed" or something? Otherwise, will need to provide translators comments here.

File: openassessment/templates/openassessmentblock/edit/oa_edit.html [hunk 2]
```diff
@@ -99,15 +99,16 @@ <h2 class="openassessment_alert_title">{% trans "Rubric Change Impacts Settings
...
+                            <option value="image" {% if file_upload_type == "image" %} selected="true" {% endif %}>{% trans "Image Display"%}</option>
+                            <option value="file" {% if file_upload_type == "file" %} selected="true" {% endif %}>{% trans "File Link"%}</option>
```
Comment (sarina) [inline openassessment/templates/openassessmentblock/edit/oa_edit.html]: both of the lines above will need translators comments.

File: openassessment/templates/openassessmentblock/peer/oa_peer_assessment.html [hunk 3]
```diff
@@ -70,15 +70,23 @@ <h3 class="peer-assessment__display__title">
                                     {{ peer_submission.answer.text|linebreaks }}
                                 </div>
 
-                                {% if allow_file_upload and peer_file_url %}
+                                {% if file_upload_type and peer_file_url %}
                                 <header class="peer-assessment__display__header">
                                     <h3 class="peer-assessment__display__title">
-                                        {% trans "Associated Image" %}
+                                        {% trans "Associated Uploaded File" %}
```
Comment (sarina) [inline openassessment/templates/openassessmentblock/peer/oa_peer_assessment.html]: This sounds awkward, maybe just, "associated file"?

File: openassessment/templates/openassessmentblock/peer/oa_peer_turbo_mode.html [hunk 4]
```diff
@@ -54,16 +54,24 @@ <h3 class="peer-assessment__display__title">{% trans "Assessment #" %}
                                 {{ peer_submission.answer.text|linebreaks }}
                             </div>
 
-                            {% if allow_file_upload and peer_file_url %}
+                            {% if file_upload_type and peer_file_url %}
                                 <header class="peer-assessment__display__header">
                                     <h3 class="peer-assessment__display__title">
                                         {% trans "Associated Image" %}
```
Comment (sarina) [inline openassessment/templates/openassessmentblock/peer/oa_peer_turbo_mode.html]: Associated File

File: openassessment/xblock/openassessmentblock.py [hunk 5]
```diff
@@ -116,10 +117,10 @@ class OpenAssessmentBlock(
         help="ISO-8601 formatted string representing the submission due date."
     )
 
-    allow_file_upload = Boolean(
-        default=False,
+    file_upload_type = String(
+        default=None,
         scope=Scope.content,
-        help="File upload allowed with submission."
+        help="File upload to be included with submission (can be 'image' or 'file')."
```
Comment (sarina) [inline openassessment/xblock/openassessmentblock.py]: why the change from "allowed" to "to be included"?

Also I think that parenthetical would benefit from a broader description - for example, "(can be None, 'image', to allow .jpg files, or 'file', to allow .pdf files)"  --- basically, to clarify what all of those options mean.

File: openassessment/xblock/static/js/spec/lms/oa_response.js [hunk 6]
```diff
@@ -401,21 +401,36 @@ describe("OpenAssessment.ResponseView", function() {
 
     it("selects too large of a file", function() {
         spyOn(baseView, 'toggleActionError').andCallThrough();
-        var files = [{type: 'image/jpg', size: 6000000, name: 'huge-picture.jpg', data: ''}];
-        view.prepareUpload(files);
+        var files = [{type: 'image/jpeg', size: 6000000, name: 'huge-picture.jpg', data: ''}];
```
Comment (sarina) [inline openassessment/xblock/static/js/spec/lms/oa_response.js]: why the change to "jpeg"? "jpg" is more common, I thought, and used later on in this file.

File: openassessment/xblock/static/js/src/lms/oa_response.js [hunk 7]
```diff
@@ -446,25 +466,32 @@ OpenAssessment.ResponseView.prototype = {
...
         this.files = null;
-        this.imageType = files[0].type;
+        this.fileType = files[0].type;
         if (files[0].size > this.MAX_FILE_SIZE) {
             this.baseView.toggleActionError(
-                'upload', gettext("File size must be 5MB or less.")
+                'upload',
+                gettext("File size must be 5MB or less.")
+            );
+        } else if (uploadType === "image" && this.UPLOADABLE_IMAGE_MIME_TYPES.indexOf(this.fileType) === -1) {
```
Comment (sarina) [inline openassessment/xblock/static/js/src/lms/oa_response.js]: please provide a comment to explain what the second check is testing.

File: openassessment/xblock/static/js/src/lms/oa_response.js [hunk 8]
```diff
@@ -446,25 +466,32 @@ OpenAssessment.ResponseView.prototype = {
...
+            this.baseView.toggleActionError(
+                'upload',
+                gettext("File must be of an allowed image type.")
             );
-        } else if (this.imageType.substring(0,6) != 'image/') {
+        } else if (uploadType === "file" && this.UPLOADABLE_FILE_MIME_TYPES.indexOf(this.fileType) === -1) {
```
Comment (sarina) [inline openassessment/xblock/static/js/src/lms/oa_response.js]: please provide a comment to explain what the second check is testing.

File: openassessment/xblock/static/js/src/lms/oa_response.js [hunk 9]
```diff
@@ -446,25 +466,32 @@ OpenAssessment.ResponseView.prototype = {
...
             this.baseView.toggleActionError(
-                'upload', gettext("File must be an image.")
+                'upload',
+                gettext("File must be of an allowed file type.")
```
Comment (sarina) [inline openassessment/xblock/static/js/src/lms/oa_response.js]: can this please speicfy what the allowed types are?

File: openassessment/xblock/static/js/src/lms/oa_response.js [hunk 10]
```diff
@@ -446,25 +466,32 @@ OpenAssessment.ResponseView.prototype = {
...
         if (files[0].size > this.MAX_FILE_SIZE) {
             this.baseView.toggleActionError(
-                'upload', gettext("File size must be 5MB or less.")
+                'upload',
+                gettext("File size must be 5MB or less.")
+            );
+        } else if (uploadType === "image" && this.UPLOADABLE_IMAGE_MIME_TYPES.indexOf(this.fileType) === -1) {
+            this.baseView.toggleActionError(
+                'upload',
+                gettext("File must be of an allowed image type.")
```
Comment (sarina) [inline openassessment/xblock/static/js/src/lms/oa_response.js]: can this please speicfy what the allowed types are?

File: openassessment/xblock/submission_mixin.py [hunk 11]
```diff
@@ -25,6 +25,21 @@ class SubmissionMixin(object):
     work outside the scope of OpenAssessmentBlock.
 
     """
+    
+    UPLOADABLE_IMAGE_MIME_TYPES = set([
+        'image/gif',
+        'image/jpeg',
```
Comment (sarina) [inline openassessment/xblock/submission_mixin.py]: will this accept a `.jpg`?

**Comment by sarina** [comment]
@caesar2164 please update the description of this PR with instructions on how to test it. Are there flags I need to set? Do you have a .tar.gz of the course that will make testing this feature easier? Etc. Please read through the [cover letter instructions](http://edx.readthedocs.org/projects/edx-developer-guide/en/latest/process/cover-letter.html) and provide as much detail as humanly possible.

**Comment by sarina** [comment]
@mhoeber ping: this will need review of strings, from doc people, and has doc impact.

File: openassessment/xblock/static/js/spec/lms/oa_response.js (hunk 6 above)
Comment (caesar2164) [inline openassessment/xblock/static/js/spec/lms/oa_response.js]: an image.jpg is of type image/jpeg. file type image.jpg is invalid...

File: openassessment/xblock/submission_mixin.py (hunk 11 above)
Comment (caesar2164) [inline openassessment/xblock/submission_mixin.py]: yes, as stated in oa_response.js an image.jpg file is still of type image/jpeg...

File: openassessment/xblock/static/js/spec/lms/oa_response.js (hunk 6 above)
Comment (sarina) [inline openassessment/xblock/static/js/spec/lms/oa_response.js]: haha ok you see how much I know about image types :grinning:  thanks for bearing with me :grimacing:

**Comment by mhoeber** [comment]
Backlogged https://openedx.atlassian.net/browse/DOC-1635

**Comment by sarina** [comment]
@caesar2164 how's this work going?

By the way you will need to rebase this - there are conflicts.

**Comment by caesar2164** [comment]
@sarina - progressing quite nicely. I've been having to back burner this every once in a while when other things take precedence...

**Comment by sarina** [comment]
@caesar2164 cool. let me know if I can do anything to help, and when this is ready to be re-reviewed.

**Comment by sarina** [comment]
Hi @caesar2164 : how's it going on this? It needs another rebase for conflicts. I'm excited to pass this off to the ORA team for review, but I don't want to do that until I know you can work on this.

**Comment by caesar2164** [comment]
@sarina - this is very close, we just need to resolve a possible S3 permission issue, and it should be good to go.

**Comment by sarina** [comment]
@caesar2164 and of course respond to my comments :)

**Comment by sarina** [comment]
Any updates on this?

**Comment by caesar2164** [comment]
Yes, @sarina, I'm GOING to work on this soon. But sadly as of right now I have a few things that jumped the queue and need attention more pressingly... :cry:

**Comment by sarina** [comment]
@caesar2164 every week at grooming I have to check on all the outstanding, open PRs. Given that, I would like to close this PR until you're ready for me to take another review pass and hand off to the ORA2 team. I can't hand it off to them until my comments are addressed, the PR is rebased, and you're ready to respond to their feedback. Do you have any objections?

**Comment by caesar2164** [comment]
No, that's fine, I'll close...
